File: src/Belajar_SOLID_09_OOP_Realization_Atau_Implementation/Java/Main.java
package Belajar_SOLID_09_OOP_Realization_Atau_Implementation.Java;

public class Main {

    interface ProductService{
        public void getProduct(String id);
        public void calculateProductPrice();
    }

    class ProductServiceImpl implements ProductService{

        @Override
        public void getProduct(String id) {

        }

        @Override
        public void calculateProductPrice() {

        }
    }
}


//    Hubungan ini menandakan bahwa classifier mengimplementasikan kontrak yang
//    telah ditentukan oleh kelas interface. Kelas classifier dapat mengimplementasikan
//    satu atau beberapa kelas interface. Classifier yang mengimplementasikan interface
//    mendukung serangkaian fitur yang dimiliki oleh kelas interface.

//    Pada contoh kode diatas kelas ProductServiceImpl merupakan kelas classifier yang
//    mengimplementasikan kontrak dari kelas interface-nya yaitu kelas ProductService. 

File: src/Belajar_SOLID_09_OOP_Realization_Atau_Implementation/Kotlin/Main.kt
package Belajar_SOLID_09_OOP_Realization_Atau_Implementation.Kotlin

interface ProductService{
    fun getProduct(id:String)
    fun calculateProductPrice()
}

class ProductServiceImpl() : ProductService{
    override fun getProduct(id: String) {
        TODO("Not yet implemented")
    }

    override fun calculateProductPrice() {
        TODO("Not yet implemented")
    }

}

// Hubungan realisasi atau implementasi adalah hubungan abstraksi
// khusus antara dua kelas, satu mewakili kelas yang
// direalisasikan (supplier) dan yang lainnya mewakili kelas implementasinya (client).
// Realisasi dapat digunakan dalam beberapa hal seperti untuk menyempurnakan sebuah
// kelas, optimasi, transformasi, template, model sintesis, komposisi kerangka kerja, dll.
// Interface realization adalah hubungan realisasi khusus antara classifier dan interface

//    Pada contoh kode diatas kelas ProductServiceImpl merupakan kelas classifier yang
//    mengimplementasikan kontrak dari kelas interface-nya yaitu kelas ProductService. 